Reformulate the release process documentation (#9736)
ComplianceAsCode/content: docs/release_process.md

This process reduces the risk of bringing a sudden breaking change into the release,
and ultimately allows the release process to happen while development continues on master branch.

## Preparing the stabilization

- Create a new branch called **stabilization-vX.Y.Z** where **X.Y.Z** is the version of upstream
release to be created.
    - For example, if the last released version was **0.1.64**, the branch should be named as
    `stabilization-v0.1.65`
<!-- -->
- Create a new Milestone called **X.Y.Z**, where **X.Y.Z** is the version of future upstream release.
    - For example, if you just created the `stabilization-v0.1.65` branch in order to start the
    stabilization phase for **0.1.65**, the new Milestone should be `0.1.66`.
<!-- -->
- All Pull Requests which are open at the moment the stabilization branch was created won't be

> included in the respective release. Therefore, if they have a Milestone defined, the Milestone

Reviewer: ```suggestion
included in this release. Therefore, if they have a Milestone defined, they
```

Author: Why is it better to change `the respective` by `this`? Could you share how did you interpret it, please?

Reviewer: To me, `respective` gives the signal that we are handling multiple releases, that each PR has a release pertaining to it. Simplifying and reducing the sentence: `Pull Requests won't be included in the respective release`. And it should probably be `their respective`. 
Ref: https://www.collinsdictionary.com/us/dictionary/english/respective

Instead of `this`, you could also use `current`.

Author: Thanks for sharing your thoughts. So, changing to `current` makes more sense to me in order to reduce the risk of misinterpretation. Anyway, I will review how to improve the complete sentence.

Author: I updated the sentence. Could you take a look again, please?

Reviewer: Looks good.

must be updated to refer the next release.
    - For example, when the `stabilization-v0.1.65` is created, all open Pull Requests with the
    Milestone `0.1.65` should be updated to the Milestone `0.1.66`.
<!-- -->
- All Issues which are open at the moment the stabilization branch was created won't be included

> in the respective release. Therefore, like Pull Requests, if they have a Milestone defined, the
> Milestone must be updated to refer the next release.

Reviewer: ```suggestion
in the this release. The same applies to Pull Requests, if they have a Milestone defined,
they must be updated to refer the next release.
```

    The release of Content version 0.1.65 is scheduled for December 2nd.
    As part of the release process, a stabilization branch was created.
    Issues and PRs that were not solved were moved to the 0.1.66 milestone.

    Any bug fixes you would like to include in release 0.1.65 should be proposed
    for the *stabilization-v0.1.65* and *master* branches as a measure to avoid
    potential conflicts between these branches.

    The next version, 0.1.66, is scheduled to be released on February 3rd,
    with the stabilization phase starting on January 23th, 2022.

    Regards,
    ```

## During the stabilization

### Bug Fixes
- Run whatever extra tests you have and report bugs as Upstream issues using the
[General Issue](https://github.com/ComplianceAsCode/content/issues/new?assignees=&labels=&template=general_issue.md) template.
    - Propose bug fixes by targeting the PRs to the **stabilization** branch. e.g. `stabilization-v0.1.65`
    - Once the PR is **merged** in the **stabilization** branch, cherry-pick it to the **master** branch.
    For example:
    ```
    git checkout master
    git pull upstream master
    git checkout -b backport_from_stabilization

>     git cherry-pick abcd1234

Reviewer: This is not clear about what commit should be cherry picked.
Moreover, I don't think cherry picking is the way to go.
If the PR had one commit we can cherry pick it. But what if it had multiple commits? I don't like much the idea of cherry-picking Merge commits to create a PR.

What I usually do is:
1. Try to create both PRs from a single branch. This is only possible when the files  in both branches have not diverged changed, or any changes in them won't cause conflicts.
```
stab_base=$(git merge-base  master stabilization-v0.1.64)
git checkout -b my-fix $stab_base
# Do the fix, commit and create both PRs through GitHub Web interface
```
2. Fix on stabilization, then rebase all the commits to a new branch from master.
```
# This assumes that stabilization-v0.1.64 is up to date,
# and branch my-fix was branched from stabilization-0.1.64 and has the fixing commits
git checkout my-fix
base_my_fix=$(git merge-base my-fix stabilization-v0.1.64)
git rebase --onto master $base_my_fix HEAD
git checkout -b my-fix-on-master
# Push my-fix and my-fix-on-master and create PRs
```

Author: There is a note below highlighting that there are different ways to do it and the cherry-picking is just "one" example. The decision on how to do it is on the author's call. That said, I am fine to include one more example for PRs with multiple commits. This second example could be useful and these two examples should be more than enough.

Reviewer: If you want to include only one example, I'd prefer that the first one be included. As it can simplify a lot the creation of PRs for master and stabilization with a single branch.

But why not include all three of them?

Author: I wrote that "I am fine to include one more example", not to keep "only one example". : )
The technical details on how do do it is on the author's freedom to decide. Since the fix is sent to both branches, it is all fine.
The examples are intended to clarify the goal and not to confuse it. So, IMO, two simple examples are pretty much enough here. Could you send a commit proposal to amend this section with your example, please? Or send it to me and I include it in the next update.

Reviewer: I can add the two examples. Should I wait for any update you may be doing in this branch?

Author: It is updated from my side.

Reviewer: @marcusburghardt I've added a few commits, please take a look.

master branch, every Friday during the stabilization phase, a PR was created to merge the
stabilization changes into the master branch. However, this approach was prone to conflicts that
could be complex to be solved after one week of asynchronous contributions on master branch.
Therefore, since the `0.1.63` release, a simpler and more reliable process has been adopted where
the authors themselves ensure that the fixes are merged into each branch at almost the same time.
This reduces the risk of conflicts and also makes it easier for authors to resolve them as the
changes are still "fresh".

### Tests

There is a GitHub Action hooked up with **stabilization** branch that will run a set of tests on every push.
Make sure that all these tests are passing before moving on with the release process.

# Release

## Updating Contributors List

- Update the contributors list by accessing the root folder of Content repository, in the
**stabilization** branch, and executing the following command:
    ```
    PYTHONPATH=. utils/generate_contributors.py`
    ```
    - De-duplicate names if necessary.

>     - Make a commit send the a PR.

Reviewer: The PR to update the contributors list needs to be sent to both master and stabilization branches.

I wonder if it makes sense to update the contributors list before stabilization is branched. :thinking:

Author: It would be more efficient to do it before the stabilization branch is created. We can definitely improve this.

    - Reference: https://github.com/ComplianceAsCode/content/pull/9601

## Checklist

- Make sure the version in `CMakeLists.txt` is correct.
    - The version should correspond to an unreleased version number.

> - Make sure you don’t have any uncommitted changes, otherwise they may be lost during the release.

Reviewer: ```
- Make sure you don’t have any uncommitted changes, otherwise they may be lost during the release.
- Make sure you have the **stabilization** branch checked out and up to date.
```
These items were part of the process to update the contributors list, it doesn't make sense in the `Checklist`.

Author: Ok, so it can also be simplified if the contributors list is updated before the stabilization branch.

The general rule is that the PR Titles will compose the body of the changelog.

### Tagging

- Create and push a **vX.Y.Z** tag to the GitHub repo.
```
git tag vX.Y.Z stabilization-vX.Y.Z
git push --tags
```
- Wait for the release action to finish.
- Check the release draft and update the release notes if necessary.

- Publish the release.

_NOTE: In case there is a need to run the job again, delete the release
draft and run the GitHb Action again._

# Clean Up

- Update the **stable** branch to point to the new release:
```
git checkout stable
git merge stabilization-vX.Y.Z

> ```

Reviewer: I'd add here that some conflicts may have to be solved.

Author: Yes, why not? : )
